### PR TITLE
Omit AZ-aware agents for non-az aware loadbalancers, use of az profil…

### PR DESCRIPTION
**Pull request by notandy**
…e metadata

Moved get_candidates from amphora_repository to it's own scheduler class.
Reworked get_candidates to support both, listener and load_balancer aware
scheduling within the same sql query.
Support az validation.
Minor cleanups
Auto-creation of AZs is now safguarded by transaction and automatically retried.

### CODE REVIEW

File: octavia_f5/api/drivers/f5_driver/driver.py [hunk 1]
```diff
@@ -234,16 +234,6 @@ def create_vip_port(self, loadbalancer_id, project_id, vip_dictionary):
...
-    def loadbalancer_failover(self, loadbalancer_id):
-        """Performs a fail over of a load balancer.
-
-        :param loadbalancer_id (string): ID of the load balancer to failover.
-        :return: Nothing if the failover request was accepted.
-        :raises DriverError: An unexpected error occurred in the driver.
-        :raises: NotImplementedError if driver does not support request.
-        """
-        raise exceptions.NotImplementedError()
-
```
Comment (BenjaminLudwigSAP) [inline octavia_f5/api/drivers/f5_driver/driver.py:246]: Why did you remove this?

File: octavia_f5/controller/worker/controller_worker.py [hunk 2]
```diff
@@ -577,36 +580,38 @@ def ensure_host_set(self, loadbalancer):
...
 
-    def register_in_availability_zone(self):
+    @oslo_db_api.wrap_db_retry(max_retries=5, retry_on_deadlock=True)
+    def register_in_availability_zone(self, az_name):
         """
         Register this worker to its assigned availability zone by creating/modifying the corresponding DB entry.
 
-        An AZ can have multiple workers (multiple F5 devices), so their names are just set as description of the AZ
-        in the DB, separated by spaces.
+        An AZ can have multiple workers (multiple F5 device-pairs), so the worker host are set in the
```
Comment (BenjaminLudwigSAP) [inline octavia_f5/controller/worker/controller_worker.py]: `s/host/hosts`

File: octavia_f5/controller/worker/controller_worker.py [hunk 3]
```diff
@@ -577,36 +580,38 @@ def ensure_host_set(self, loadbalancer):
                                            id=loadbalancer.id,
                                            server_group_id=CONF.host[:36])
 
-    def register_in_availability_zone(self):
+    @oslo_db_api.wrap_db_retry(max_retries=5, retry_on_deadlock=True)
+    def register_in_availability_zone(self, az_name):
         """
         Register this worker to its assigned availability zone by creating/modifying the corresponding DB entry.
```
Comment (BenjaminLudwigSAP) [inline octavia_f5/controller/worker/controller_worker.py]: Please adjust docstring according to method's signature: `s/its assigned/an`

File: octavia_f5/controller/worker/controller_worker.py [hunk 4]
```diff
@@ -577,36 +580,38 @@ def ensure_host_set(self, loadbalancer):
...
-                        'name': az_name,
-                        'description': CONF.host,
-                        'enabled': True,
-                        # AZ profile 00000000-0000-0000-0000-000000000000 is always present, so we just use that
-                        'availability_zone_profile_id': '00000000-0000-0000-0000-000000000000'
-                    }
-                    self._az_repo.create(session, **az_dict)
\ No newline at end of file
+        with db_apis.get_lock_session() as lock_session:
+            az = self._az_repo.get(lock_session, name=az_name)
```
Comment (BenjaminLudwigSAP) [inline octavia_f5/controller/worker/controller_worker.py:592]: Please move this line down above `if az:`

File: octavia_f5/controller/worker/controller_worker.py [hunk 5]
```diff
@@ -577,36 +580,38 @@ def ensure_host_set(self, loadbalancer):
...
+            metadata = self._az_repo.get_availability_zone_metadata_dict(lock_session, az_name)
+            hosts = metadata.get('hosts', [])
+            if az:
+                if not CONF.host in hosts:
+                    # add host to availibility zone metadata (profile)
```
Comment (BenjaminLudwigSAP) [inline octavia_f5/controller/worker/controller_worker.py]: `s/metadata (profile)/profile metadata`

File: octavia_f5/db/scheduler.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+LOG = logging.getLogger(__name__)
+
+
+class Scheduler(object):
+    def __init__(self):
+        self.az_repo = repositories.AvailabilityZoneRepository()
+
+    def get_candidates(self, session, az_name=None):
+        """ Get F5 (active) BigIP host candidate depending on the load (amount of listeners in amphora vrrp_priority
+        column) and the desired availability zone.
```
Comment (BenjaminLudwigSAP) [inline octavia_f5/db/scheduler.py]: `s/)/ or amount of load balancers)/`

File: octavia_f5/controller/worker/controller_worker.py [hunk 7]
```diff
@@ -577,36 +580,38 @@ def ensure_host_set(self, loadbalancer):
...
-                        'description': CONF.host,
-                        'enabled': True,
-                        # AZ profile 00000000-0000-0000-0000-000000000000 is always present, so we just use that
-                        'availability_zone_profile_id': '00000000-0000-0000-0000-000000000000'
-                    }
-                    self._az_repo.create(session, **az_dict)
\ No newline at end of file
+        with db_apis.get_lock_session() as lock_session:
+            az = self._az_repo.get(lock_session, name=az_name)
+            metadata = self._az_repo.get_availability_zone_metadata_dict(lock_session, az_name)
```
Comment (BenjaminLudwigSAP) [inline octavia_f5/controller/worker/controller_worker.py]: This fails at first startup, because there is no AZ profile DB entry present yet. This raises `sqlalchemy.orm.exc.NoResultFound`, which we need to catch and then create an AZ profile.

**Review by BenjaminLudwigSAP — CHANGES_REQUESTED** [review]
Nice PR, I especially like your refactoring of the sqlalchemy mess I produced :)
Only real problem is that AZ profile entries already need to be existing in the DB.

File: octavia_f5/api/drivers/f5_driver/driver.py (hunk 1 above)
Comment (notandy) [reply octavia_f5/api/drivers/f5_driver/driver.py:246]: It's been duplicated, see line 80

File: octavia_f5/controller/worker/controller_worker.py (hunk 3 above)
Comment (notandy) [reply octavia_f5/controller/worker/controller_worker.py]: Thx!

File: octavia_f5/controller/worker/controller_worker.py (hunk 2 above)
Comment (notandy) [reply octavia_f5/controller/worker/controller_worker.py]: :+1

File: octavia_f5/controller/worker/controller_worker.py (hunk 4 above)
Comment (notandy) [reply octavia_f5/controller/worker/controller_worker.py:592]: We need the lock_session to access the az_repo

File: octavia_f5/controller/worker/controller_worker.py (hunk 7 above)
Comment (notandy) [reply octavia_f5/controller/worker/controller_worker.py]: good catch!

File: octavia_f5/db/scheduler.py (hunk 6 above)
Comment (notandy) [reply octavia_f5/db/scheduler.py]: thx

File: octavia_f5/controller/worker/controller_worker.py (hunk 5 above)
Comment (notandy) [reply octavia_f5/controller/worker/controller_worker.py]: thx

File: octavia_f5/controller/worker/controller_worker.py (hunk 4 above)
Comment (BenjaminLudwigSAP) [reply octavia_f5/controller/worker/controller_worker.py:592]: I'm talking about the `az = (...)` line. We don't need `az` before `if az:`.

File: octavia_f5/controller/worker/controller_worker.py (hunk 7 above)
Comment (BenjaminLudwigSAP) [reply octavia_f5/controller/worker/controller_worker.py]: Unfortunately we can't reuse `lock_session` in the except-block, else I would already have made a quickfix.

Comment (notandy) [reply octavia_f5/controller/worker/controller_worker.py]: I fixed it by moving the get_availability_zone_metadata_dict inside the if az block, which should fix the issue since no AZ is created without an related AZ-Profile

File: octavia_f5/db/scheduler.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+        :param session: A Sql Alchemy database session.
+        :param az_name: Name of the availability zone to schedule to. If it is None, all F5 amphora are considered.
+        """
+
+        # get all hosts
+        candidates = session.query(
+            models.Amphora.compute_flavor,
+            func.count(models.LoadBalancer.id)
+        ).join(
```
Comment (BenjaminLudwigSAP) [inline octavia_f5/db/scheduler.py:43]: Note: In my PR I explicitely decided against a join, because it was quite slow in terms of performance. But that may have been due to a poorly formulated SQL query, I don't remember what exactly it was. I will measure performance for this query in QA1 after upgrade. We may want to do that in production as well after the first rollouts.

File: octavia_f5/db/scheduler.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        else:
+            # we need to filter out all az-aware hosts
+            azs = set([az.name for az in self.az_repo.get_all(session)[0]])
+            omit_hosts = set()
+            for az in azs:
+                metadata = self.az_repo.get_availability_zone_metadata_dict(session, az)
+                omit_hosts.update(metadata.get('hosts', []))
+            candidates = candidates.filter(
+                models.Amphora.compute_flavor.notin_(omit_hosts))
+        return [candidate[0] for candidate in candidates.all()]
```
Comment (BenjaminLudwigSAP) [inline octavia_f5/db/scheduler.py:82]: We need error handling here in case the list is empty.
Either `get_candidates` handles that case itself by logging an error and raising an exception (I would prefer that, since the semantics are that the caller can expect *some* candidates to be returned by this function).
Or the caller checks whether candidates are present and otherwise logs an error and raises an exception. (I would not prefer this, since that would necessitate boilerplate code for every caller we write in the future)
In any case the docstring of `get_candidates` has to be updated as well.

However I can add this in my PR after this one has been merged, that's fine.

**Review by BenjaminLudwigSAP — APPROVED** [review]
Just one thing, but I can fix that in my own PR after merging this one.

File: octavia_f5/db/scheduler.py (hunk 9 above)
Comment (notandy) [reply octavia_f5/db/scheduler.py:82]: get_candidates gives back a list of candidates, and returning an empty list is handled by the Caller already.
